File: src/class068/C3_InterleavingString.py
class Code03_InterleavingString:
    # 交错字符串
    # 给定三个字符串 s1、s2、s3
    # 请帮忙验证s3是否由s1和s2交错组成
    @staticmethod
    def is_interleave1(s1, s2, s3):
        if len(s1) + len(s2) != len(s3):
            return False
        n, m = len(s1), len(s2)
        dp = [[False] * (m + 1) for _ in range(n + 1)]
        dp[0][0] = True
        
        for i in range(1, n + 1):
            if s1[i - 1] != s3[i - 1]:
                break
            dp[i][0] = True
        
        for j in range(1, m + 1):
            if s2[j - 1] != s3[j - 1]:
                break
            dp[0][j] = True
        
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                dp[i][j] = (s1[i - 1] == s3[i + j - 1] and dp[i - 1][j]) or (s2[j - 1] == s3[i + j - 1] and dp[i][j - 1])
        
        return dp[n][m]

    # 空间压缩
    @staticmethod
    def is_interleave2(s1, s2, s3):
        if len(s1) + len(s2) != len(s3):
            return False
        n, m = len(s1), len(s2)
        dp = [False] * (m + 1)
        dp[0] = True
        
        for j in range(1, m + 1):
            if s2[j - 1] != s3[j - 1]:
                break
            dp[j] = True
        
        for i in range(1, n + 1):
            dp[0] = s1[i - 1] == s3[i - 1] and dp[0]
            for j in range(1, m + 1):
                dp[j] = (s1[i - 1] == s3[i + j - 1] and dp[j]) or (s2[j - 1] == s3[i + j - 1] and dp[j - 1])
        
        return dp[m]

File: src/class068/C3_InterleavingString.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    // Dynamic programming solution
    bool isInterleave1(string str1, string str2, string str3) {
        if (str1.length() + str2.length() != str3.length()) {
            return false;
        }
        int n = str1.length();
        int m = str2.length();
        vector<vector<bool>> dp(n + 1, vector<bool>(m + 1, false));
        dp[0][0] = true;
        for (int i = 1; i <= n; ++i) {
            dp[i][0] = (str1[i - 1] == str3[i - 1]) && dp[i - 1][0];
        }
        for (int j = 1; j <= m; ++j) {
            dp[0][j] = (str2[j - 1] == str3[j - 1]) && dp[0][j - 1];
        }
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                dp[i][j] = ((str1[i - 1] == str3[i + j - 1]) && dp[i - 1][j]) || ((str2[j - 1] == str3[i + j - 1]) && dp[i][j - 1]);
            }
        }
        return dp[n][m];
    }

    // Space optimized dynamic programming solution
    bool isInterleave2(string str1, string str2, string str3) {
        if (str1.length() + str2.length() != str3.length()) {
            return false;
        }
        int n = str1.length();
        int m = str2.length();
        vector<bool> dp(m + 1, false);
        dp[0] = true;
        for (int j = 1; j <= m; ++j) {
            dp[j] = (str2[j - 1] == str3[j - 1]) && dp[j - 1];
        }
        for (int i = 1; i <= n; ++i) {
            dp[0] = (str1[i - 1] == str3[i - 1]) && dp[0];
            for (int j = 1; j <= m; ++j) {
                dp[j] = ((str1[i - 1] == str3[i + j - 1]) && dp[j]) || ((str2[j - 1] == str3[i + j - 1]) && dp[j - 1]);
            }
        }
        return dp[m];
    }
};
